### ISSUE TITLE
What should the server do if a user submits a Thing Description with  `POST` request on `/things` which contains an `id` in the body?

### ISSUE THREAD

**Issue by benfrancis** [w3c/wot-discovery#381]
Currently [Zion](https://github.com/vaimee/zion)'s API documentation says that "The server rejects the request if there is an id in the body". I've found that in practice this isn't true and it will actually accept an ID in the body, but it returns a different ID (a system-generated URN) in the Location header of the 201 Created response. This is bad because the URI in the Thing Description resource URL doesn't match the ID in the Thing Description resource. A side effect of this is that if you request a list of Thing Descriptions there's no way to find out the URL of the individual Thing Description resource from the response.

I noticed that the specification doesn't actually say what to do in this situation, it just says to validate the TD before creating it, but a TD containing an `id` would be valid.

Perhaps the specification should say that submitting a `POST` request with an `id` in the body should generate an error response?

Edit: Second question moved to #382

P.S. Thanks for sharing Vaimee's early implementation, @relu91. This shows why implementation experience make such a difference. (I'm also currently implementing the specification and grappling with the same issues.)

**Comment by benfrancis** [comment]
For the record, currently if you add a web thing with an HTTP URL as an ID to WebThings Gateway, it gets given a URL like:

`http://localhost:8080/things/https---mywebthingserver.com-things-lamp` (with all the non-alphanumeric characters replaced with dashes)

Since WebThings Gateway acts as a gateway, not just a directory, it also generates a new Thing Description for that Thing, using that URL as the new `id` member.

**Comment by farshidtz** [comment]
Could you please separate the questions into two issues? Otherwise it will be hard to discuss and resolve them.

> What should the server do if a user submits a Thing Description with  `POST` request on `/things` which contains an `id` in the body?

I believe we originally had a statement in the spec (or PR) to reject such operations with a 409 Conflict status. But it got removed in the discussions around allowing distinct resource and TD ids. 

I think it makes total sense to have that back.

**Comment by benfrancis** [comment]
> Could you please separate the questions into two issues?

Yes sorry, I've split the second question into #382.

**Comment by relu91** [comment]
We had a long discussion about this some days ago and we concluded that it was indeed a bug in Zion. We had also the fix available but somehow it didn't make it to the alpha version cause we didn't have a unit test to verify it. 

Our reasoning was the following. If check the assertions under the creation section you have:
> An [Anonymous TD](https://w3c.github.io/wot-discovery/#dfn-wot-anonymous-thing-description) MUST be submitted to the directory in the body of an HTTP POST request at /things endpoint.

The assertion clearly states the type of payload of the endpoint: an Anonymous TD. As you know an Anonymous TD by convention it is a TD without an ID, so practically the endpoint should refuse to inject something other than a TD without an id. 

I hope this explains why we had a misleading message in the swagger 😄 , nonetheless I would not complain if the spec was a little bit more explicit on this. In the "unpublished" branch we returned Bad Request as if the input was not valid, but also Conflict Status sounds like a good error message. 

> P.S. Thanks for sharing Vaimee's early implementation, @relu91. This shows why implementation experience make such a difference. (I'm also currently implementing the specification and grappling with the same issues.)

Thank you @benfrancis for giving us early feedback! It takes time to test and write issues so I really appreciated it.
